pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,508 @@
+			GatewayIP:       key,
+			LastMeasuredRTT: entry.LastMeasuredRTT.Microseconds(),
+			LastSendTime:    entry.LastSendTime.UnixMicro(),
+			LastRecvTime:    entry.LastRecvTime.UnixMicro(),
+		})
+	}
+
+	// Report to controller
+	antreaClient, err := m.antreaClientProvider.GetAntreaClient()
+	if err != nil {
+		klog.ErrorS(err, "Failed to get Antrea client")
+		return
+	}
+	_, err = antreaClient.ControlplaneV1beta2().NodeIPLatencyStats().Create(context.TODO(), nodeIPLatencyStats, metav1.CreateOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Failed to report latency data")
+		return
+	}
+}
+
+func (m *NodeLatencyMonitor) Run(stopCh <-chan struct{}) {
+	// Top level goroutine to handle termination
+	go func() {
+		<-stopCh
+	}()
+
+	// Start the monitor loop
+	go m.nodeLatencyMonitorInformer.Informer().Run(stopCh)
+	go m.nodeInformer.Informer().Run(stopCh)
+	go m.monitorLoop(stopCh)
+}
+
+func (m *NodeLatencyMonitor) monitorLoop(stopCh <-chan struct{}) {
+	// Low level goroutine to handle ping loop
+	var currentTicker *time.Ticker
+	var tickerCh <-chan time.Time
+	var ipv4Socket, ipv6Socket net.PacketConn
+	var err error
+	tickerStopCh := make(chan struct{})
+
+	defer func() {
+		if ipv4Socket != nil {
+			ipv4Socket.Close()
+		}
+		if ipv6Socket != nil {
+			ipv6Socket.Close()
+		}
+		if currentTicker != nil {
+			currentTicker.Stop()
+		}
+		close(tickerStopCh)
+	}()
+
+	// Update current ticker based on the latencyConfig
+	updateTicker := func(interval time.Duration) {
+		if currentTicker != nil {
+			currentTicker.Stop() // Stop the current ticker
+		}
+		currentTicker = time.NewTicker(interval)
+		tickerCh = currentTicker.C
+	}
+
+	// Start the pingAll goroutine
+	for {
+		select {
+		case <-tickerCh:
+			// Try to send pingAll signal
+			m.pingAll(ipv4Socket, ipv6Socket)
+			// Report to controller
+			m.report()
+		case <-tickerStopCh:
+			// Close current sockets
+			if ipv4Socket != nil {
+				ipv4Socket.Close()
+			}
+			if ipv6Socket != nil {
+				ipv6Socket.Close()
+			}
+		case <-stopCh:
+			// Stop the ticker loop
+			if currentTicker != nil {
+				currentTicker.Stop()
+			}
+			close(tickerStopCh)
+			return
+		case <-m.latencyConfigChanged:
+			// Start or stop the pingAll goroutine based on the latencyConfig
+			if m.latencyConfig.Enable {
+				// latencyConfig changed
+				updateTicker(m.latencyConfig.Interval)
+
+				// If the recvPing socket is closed, restart it
+				// In case of IPv4-only or IPv6-only, we need to check the socket status,
+				// and restart it if it is closed(CRD is deleted).
+				if ipv4Socket == nil && m.gatewayConfig.IPv4 != nil {
+					// Create a new socket for IPv4 when the gatewayConfig is IPv4-only
+					ipv4Socket, err = icmp.ListenPacket(IPv4ProtocolICMPRaw, "0.0.0.0")
+					if err != nil {
+						klog.ErrorS(err, "Failed to create ICMP socket for IPv4")
+						return
+					}
+					go m.recvPing(ipv4Socket, true, tickerStopCh)
+				}
+				if ipv6Socket == nil && m.gatewayConfig.IPv6 != nil {
```

Comment:
I don't understand the logic here. In network-policy only mode, `m.gatewayConfig.IPv4` will be `nil`, because in that mode we do not assign an IP to the gateway interface. So this here is the wrong check to determine whether the cluster is IPv4-only, IPv6-only, or dual-stack.
We have the `IsIPv4Enabled` and `IsIPv6Enabled` functions defined in `pkg/agent/config` for this purpose. They take into account the traffic mode.
You will need to replace the `gatewayConfig *config.GatewayConfig` and `isNetworkPolicyOnly bool` parameters for `NewNodeLatanceyMonitor` with `nodeConfig *configNodeConfig` and `trafficEncapMode config.TrafficEncapModeType`.

Reply:
Thanks, this is indeed a big problem and I will fix it.